netbox/core/api/views.py

```diff
@@ -37,6 +37,11 @@ def sync(self, request, pk):
             raise PermissionDenied("Syncing data sources requires the core.sync_datasource permission.")
 
         datasource = get_object_or_404(DataSource, pk=pk)
+
+        # have to check perms again against this specific object as there could be constraints
+        if not request.user.has_perm('core.sync_datasource', datasource):
```

Comment:
The call to `request.user.has_perm()` above should be moved down and extended by specifying the object being synced (e.g. `has_perm('core.sync_datasource', obj=datasource)`) to properly utilize the permissions backend and avoid the need for a custom error message.